VerilogCodingStyle.md

```diff
@@ -2291,6 +2292,109 @@ always_comb begin
 end
 ```
 
+### Functions and Tasks
+
+The following section applies to synthesizable RTL only.
+
+***In synthesizable RTL use of functions is allowed, provided they are declared
+automatic. Tasks should not be used.***
+
+Functions must be declared outside of a class in either a package or inside a
+module. A package is appropriate where the function relates to other definitions
+in the package and could be useful to multiple modules (even if its currently
+only used by one). A module is appropriate where the function specifically
+relates to the internals of that module.
+
+Functions should aim to conceptually represent a reusable block of combinational
+logic.
+
+Storage types must be explicitly declared for all arguments and the function
+return value.
+
+Do not use `input`, `output`, `inout`, or `ref` on function arguments.
+
+Do not assign to function arguments in the function body.
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+ return a ^ b;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+function automatic [2:0] foo(input [2:0] a, [2:0] b);
+  b = b + 1;
+  return a + b;
+endfunction
+```
+
+It is recommended that data is returned from a function using an explicit
+`return` style. Avoid using a `function_name = result` style other than for
+simple single-expression / no control flow function.
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+ foo = a ^ b;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+function automatic [2:0] foo(input [2:0] a, [2:0] b);
+  if (a == 3'd2) begin
+    foo = b;
+  end else begin
+    foo = a ^ b;
+  end
+endfunction
+```
+
+For any local variables ensure they will get assigned on all code-paths, whether
+through an initial assignment or through the use of `else` and `default:` for
+`if` and `case` statements
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+  logic [2:0] local_var_1 = 3'd0;
```

Comment:
Was wondering if this initial assignment along with definition is ok, couldn't immediately find anything in the LRM that suggests it would be problematic.

Reply:
I think it _should_ be ok, lint is fine with it as well. In the past I always assigned defaults explicitly on a new line, though, to be on the safe side... maybe someone else has more experience with this variant?

Reply:
It's different only for static functions (without `automatic`). 

To keep people from tripping over the differences, having the initialization in a separate line gives code that works the same in static and automatic functions. So I'd write:

```
logic [2:0] local_var_1;
local_var_1 = 3'd0;
```

Reply:
agreeing with @imphil